File: internal/editors/jog.py
# jog — a memory for your working tree. Installed by `jog editors install`;
# `jog editors uninstall sublime` removes it.
#
# Snapshots the repo after every save by running `jog editor-hook sublime
# <file>`. The hook exits in milliseconds outside a git repo.
# on_post_save_async runs on Sublime's worker thread and the process is
# fully detached with every fd on DEVNULL, so saves never wait on jog —
# and a missing or moved jog is swallowed whole. The path below is baked
# by the installer because GUI editors often launch without the shell's
# PATH.

import subprocess

import sublime_plugin

JOG = "{{JOG}}"


class JogListener(sublime_plugin.EventListener):
    def on_post_save_async(self, view):
        path = view.file_name()
        if not path:
            return
        try:
            subprocess.Popen(
                [JOG, "editor-hook", "sublime", path],
                stdin=subprocess.DEVNULL,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                start_new_session=True,
            )
        except OSError:
            pass

File: internal/editors/jog.lua
-- jog — a memory for your working tree. Installed by `jog editors install`;
-- `jog editors uninstall micro` removes it.
--
-- Snapshots the repo after every save by running `jog editor-hook micro
-- <file>`. The hook exits in milliseconds outside a git repo. JobSpawn
-- is micro's async runner; with nil callbacks the output vanishes and a
-- missing jog fails without a trace.

VERSION = "1.0.0"

local shell = import("micro/shell")

function onSave(bp)
    local path = bp.Buf.AbsPath
    if path ~= nil and path ~= "" then
        shell.JobSpawn("jog", {"editor-hook", "micro", path}, nil, nil, nil)
    end
    return false
end
